### PR TITLE
refactored to use parser/renderer. 

### PR DESCRIPTION
**Pull request by jonschlinkert**
@es128 / @shinnn / @doowb if any of you have time to review, any feedback would be appreciated.

This code is much easier to maintain and debug. There are [some failing unit tests](https://github.com/jonschlinkert/extglob/blob/refactor/test.js#L321-L331), but IMHO it seems like some of them should be failing...  any thoughts on those (or insight into where I can find a detailed extglob spec?)

@es128 once you're good with this we can bump the minor and publish, but we should probably wait until I finish refactoring micromatch before we merge over there.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,8 +1,15 @@
 sudo: false
 language: node_js
 node_js:
-  - "0.10"
-  - "0.12"
-  - "iojs"
+  - stable
```
Comment (shinnn) [inline .travis.yml:4]: `stable` is deprecated. https://github.com/travis-ci/docs-travis-ci-com/pull/534

Comment (jonschlinkert) [inline .travis.yml:4]: thanks I keep forgetting to remove that from the `.travis.yml` template. done!

**Comment by doowb** [comment]
Looks good. I like this format better and agree that it will be better to maintain and debug.

**Comment by es128** [comment]
LGTM 👍

**Comment by jonschlinkert** [comment]
great thanks, I'll merge today

**Comment by jonschlinkert** [comment]
I'm working on optimizations to this. I'll merge as soon as that's done

**Comment by tunnckoCore** [comment]
Awesome! Benchmark is pretty important. Actually, it's interesting `snapdragon` to have some. I'm curious because started one interesting project https://github.com/limonjs/limon - on per character basis pluggable lexer, not parser, but lexer with which you can build parsers. It would be great if you have time to try it and give some feedback. :)

**Comment by jonschlinkert** [comment]
> Benchmark is pretty important.

Here are the latest benchmarks:

```
Benchmarking: (4 of 4)
 · negation-simple
 · range-false
 · range-true
 · star-simple

# benchmark/fixtures/isMatch/negation-simple.js (43 bytes)
  extglob.isMatch x 5,354,772 ops/sec ±1.03% (89 runs sampled)
  minimatch x 140,497 ops/sec ±1.46% (86 runs sampled)

  fastest is extglob.isMatch

# benchmark/fixtures/isMatch/range-false.js (56 bytes)
  extglob.isMatch x 5,445,359 ops/sec ±1.11% (87 runs sampled)
  minimatch x 118,570 ops/sec ±1.10% (89 runs sampled)

  fastest is extglob.isMatch

# benchmark/fixtures/isMatch/range-true.js (56 bytes)
  extglob.isMatch x 5,579,827 ops/sec ±1.26% (88 runs sampled)
  minimatch x 128,965 ops/sec ±1.20% (88 runs sampled)

  fastest is extglob.isMatch

# benchmark/fixtures/isMatch/star-simple.js (45 bytes)
  extglob.isMatch x 5,186,613 ops/sec ±1.08% (86 runs sampled)
  minimatch x 173,487 ops/sec ±1.09% (89 runs sampled)

  fastest is extglob.isMatch
```

**Comment by jonschlinkert** [comment]
see https://github.com/jonschlinkert/extglob/pull/5

**Pull request closed without merging**